File: C#/DSA/Easy/Trees/maxDepth.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Tree
{
    public int MaxDepth(TreeNode root)
    {
        if (root is null) return 0;

        int maxDepth = 1;
        void Traverse(TreeNode curr, int depth)
        {
            maxDepth = Math.Max(maxDepth, depth);
            if (curr.left is not null) Traverse(curr.left, depth + 1);
            if (curr.right is not null) Traverse(curr.right, depth + 1);
        }
        Traverse(root, maxDepth);
        return maxDepth;
    }

    //Bottom-up approach
    public int MaxDepth2(TreeNode root)
    {
        if (root is null) return 0;

        int Traverse(TreeNode curr)
        {
            if (curr is null) return 0;
            int leftMax = Traverse(curr.left);
            int rightMax = Traverse(curr.right);

            return 1 + Math.Max(leftMax, rightMax);
        }
        return Traverse(root);
    }
}

File: JS/Medium/BinaryTree/LevelOrderTraversal/maxDepth.js
/** 104
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  //Do level order traversal
  let maxLevel = 0;
  if (!root) return maxLevel;

  let traverse = (curr, level) => {
    maxLevel = Math.max(maxLevel, level);
    curr.left && traverse(curr.left, level + 1);
    curr.right && traverse(curr.right, level + 1);
  };

  traverse(root, 0);
  return maxLevel;
};

var maxDepth = function (root) {
  //Do level order traversal
  let maxLevel = 0;
  if (!root) return maxLevel;

  let traverse = (curr, level) => {
    maxLevel = Math.max(maxLevel, level);
    curr.left && traverse(curr.left, level + 1);
    curr.right && traverse(curr.right, level + 1);
  };

  traverse(root, 0);
  return maxLevel;
};
